wkhtmltopdf/utils.py

```diff
@@ -113,58 +113,81 @@ def convert_to_pdf(filename, header_filename=None, footer_filename=None, cmd_opt
+                prefix='wkhtmltopdf', suffix='.html',
+                delete=(not debug)
+            )
+            self.filename = self.temporary_file.name
+        except:
+            # In case something fails, return an empty filename string.
+            self.filename = ''
+
+    def __del__(self):
+        # Always close the temporary_file on object destruction.
+        if self.temporary_file is not None:
+            self.temporary_file.close()
 
 def render_pdf_from_template(input_template, header_template, footer_template, context, request=None, cmd_options=None):
-    debug = getattr(settings, 'WKHTMLTOPDF_DEBUG', settings.DEBUG)
+    # For basic usage. Performs all the actions necessary to create a single
+    # page PDF from a single template and context.
     cmd_options = cmd_options if cmd_options else {}
 
-    input_file = header_file = footer_file = None
     header_filename = footer_filename = None
 
-    try:
-        input_file = render_to_temporary_file(
-            template=input_template,
+    # Main content.
+    input_file = RenderedFile(
+        template=input_template,
+        context=context,
+        request=request
+    )
+
+    # Optional. For header template argument.
+    if header_template:
+        header_file = RenderedFile(
+            template=header_template,
             context=context,
-            request=request,
-            prefix='wkhtmltopdf', suffix='.html',
-            delete=(not debug)
+            request=request
         )
+        header_filename = header_file.filename
 
-        if header_template:
-            header_file = render_to_temporary_file(
-                template=header_template,
-                context=context,
-                request=request,
-                prefix='wkhtmltopdf', suffix='.html',
-                delete=(not debug)
-            )
-            header_filename = header_file.name
+    # Optional. For footer template argument.
+    if footer_template:
+        footer_file = RenderedFile(
+            template=footer_template,
+            context=context,
+            request=request
+        )
+        footer_filename = footer_file.filename
 
-        if footer_template:
-            footer_file = render_to_temporary_file(
-                template=footer_template,
-                context=context,
-                request=request,
-                prefix='wkhtmltopdf', suffix='.html',
-                delete=(not debug)
-            )
-            footer_filename = footer_file.name
-
-        return convert_to_pdf(filename=input_file.name,
-                              header_filename=header_filename,
-                              footer_filename=footer_filename,
-                              cmd_options=cmd_options)
-    finally:
-        # Clean up temporary files
-        for f in filter(None, (input_file, header_file, footer_file)):
-            f.close()
+    return convert_to_pdf(filename=[input_file.filename],
```

Comment:
True, the list **_is_** optional. I just wanted to start showing that list arguments are possible. And that the final argument sent to wkhtmltopdf will be a list. 

But functionally, it will work either way. I will remove the listing (listering?) here if you like.